pkg/virt-dhcp/dnsmasq.go

```diff
@@ -0,0 +1,240 @@
+package virtdhcp
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"os"
+	"os/exec"
+	"sort"
+	"strconv"
+	"strings"
+	"text/template"
+
+	"syscall"
+
+	"kubevirt.io/kubevirt/pkg/logging"
+)
+
+const (
+	DNSmasqExec   = "/usr/sbin/dnsmasq"
+	DHCPConfig    = "/tmp/dhcp.conf"
+	DHCPLeaseFile = "/tmp/dhcp.leases"
+	DHCPPidFile   = "/tmp/dhcp.pid"
+)
+
+type DNSmasqInstance struct {
+	hosts    map[string]string
+	ipRange  [][]byte
+	confpath string
+	args     []string
+	cmd      *exec.Cmd
+}
+
+func NewDNSmasq() *DNSmasqInstance {
+	dnsMasq := new(DNSmasqInstance)
+	dnsMasq.hosts = make(map[string]string)
+	dnsMasq.ipRange = make([][]byte, 2)
+	dnsMasq.args = []string{"-k", "-d", "--conf-file=" + DHCPConfig, "--dhcp-leasefile=" + DHCPLeaseFile, "--pid-file=" + DHCPPidFile}
+	return dnsMasq
+}
+
+// implement `Interface` in sort package.
+type sortByteArrays [][]byte
+
+func (b sortByteArrays) Len() int {
+	return len(b)
+}
+
+func (b sortByteArrays) Less(i, j int) bool {
+	// bytes package already implements Comparable for []byte.
+	switch bytes.Compare(b[i], b[j]) {
+	case -1:
+		return true
+	case 0, 1:
+		return false
+	default:
+		return false
+	}
+}
+
+func (b sortByteArrays) Swap(i, j int) {
+	b[j], b[i] = b[i], b[j]
+}
+
+func (dnsmasq *DNSmasqInstance) Start() error {
+	logging.DefaultLogger().Debug().Msg("starting Dnsmasq")
+	cmd := exec.Command(DNSmasqExec, dnsmasq.args...)
+
+	dnsmasq.cmd = cmd
+	err := cmd.Start()
+	if err != nil {
+		logging.DefaultLogger().Error().Reason(err).Msg("failed to start dnsmasq")
+		return err
+	}
+	go func(cmd *exec.Cmd) {
+		err := cmd.Wait()
+		if err != nil {
+			logging.DefaultLogger().Warning().Reason(err).Msgf("dnsmasq process exit")
+		}
+	}(cmd)
+	return nil
+}
+
+func checkProcessExist(pid int) (bool, *os.Process) {
+	proc, err := os.FindProcess(pid)
+	if err != nil {
+		logging.DefaultLogger().Warning().Msgf("failed to find process %d", pid)
+		return false, nil
+	}
+	procErr := proc.Signal(syscall.Signal(0))
+	return procErr == nil, proc
+}
+
+func (dnsmasq *DNSmasqInstance) stop() error {
+	logging.DefaultLogger().Info().Msg("Stopping Dnsmasq")
+	if dnsmasq.cmd != nil {
+		procExist, proc := checkProcessExist(dnsmasq.cmd.Process.Pid)
```

Comment:
I have the feeling, that all this would be much simpler, if virt-dhcp would just exit too if the monitored dnsmasq process goes away, but I am not completely sure. If we do sighups to reload the config it should be fairly simple. Something like:

```golang
done := make(chan error)

cmd.Start()
go func() {
  cmd.Wait()
  close(done)
}()
go func() {
//do our stuff; reload config; ..
defer syscall.kill(...)
for {

}
}()

// If the dhcp server exits, we exit too
<- done
```

It is still possible that the application would not be able to properly stop dnsmasq (e.g. sigkilled itself). I would just read the pid file from dnsmasq then, and kill the process on restart, so that we ever only have to deal with one dnsmasq instance.

Reply:
IIRC the main issue with SIGHUP was that dnsmasq starts as root but changes to another user.
I think we can run it with --user=root all the time for the sighup to work.